From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Mon, 28 Feb 2022 08:56:16 -0600
Subject: [PATCH] Add provided Material to getDrops


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 3fc35a559ea53f045f2f72b272bc8ba34d1b2994..7fbd9c223bf34feb35f7836b42a3bd220c4cbbc5 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -624,6 +624,20 @@ public interface Block extends Metadatable, net.kyori.adventure.translation.Tran
     @NotNull
     Collection<ItemStack> getDrops(@NotNull ItemStack tool, @Nullable Entity entity);
 
+    // Slice start
+    /**
+     * Returns a list of items which would drop by the entity destroying this
+     * block as though it were a given Material with a specific tool
+     *
+     * @param blockType The block type to use as the source loot
+     * @param tool The tool or item in hand used for digging
+     * @param entity the entity destroying the block
+     * @return a list of dropped items for this type of block
+     */
+    @NotNull
+    Collection<ItemStack> getDrops(@NotNull Material blockType, @NotNull ItemStack tool, @Nullable Entity entity);
+    // Slice end
+
     /**
      * Returns if the given item is a preferred choice to break this Block.
      *
