From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Mon, 10 Apr 2023 08:46:54 -0500
Subject: [PATCH] AntiXray Bypass


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 895e1b57709edfbed7db3ed2d9e49baee663dbcf..1897af7e036483932924465719b6ab602be72a7f 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2761,6 +2761,10 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     boolean hasResourcePack();
 
+    // Slice start
+    void toggleAntiXrayBypass(boolean bypass);
+    // Slice end
+
     /**
      * Gets a copy of this players profile
      * @return The players profile object
