From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Mon, 10 Apr 2023 08:46:54 -0500
Subject: [PATCH] AntiXray Bypass


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index bff6dcf37d451e1521737326d255082cb0e32ad6..c84fbb2e325e019b20fbfae8beb2235016479fda 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2804,6 +2804,10 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     boolean hasResourcePack();
 
+    // Slice start
+    void toggleAntiXrayBypass(boolean bypass);
+    // Slice end
+
     /**
      * Gets a copy of this players profile
      * @return The players profile object
