From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Mon, 10 Apr 2023 08:46:54 -0500
Subject: [PATCH] AntiXray Bypass


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 53ac8be977419763206339b2c6ff64d8753f8bcb..e4a7771944fc9ec199764e89a7d05437a206fe4c 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2836,6 +2836,10 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     boolean hasResourcePack();
 
+    // Slice start
+    void toggleAntiXrayBypass(boolean bypass);
+    // Slice end
+
     /**
      * Gets a copy of this players profile
      * @return The players profile object
