From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Tue, 25 Apr 2023 08:16:51 -0500
Subject: [PATCH] Add Force Crit to PlayerPreAttackEntityEvent


diff --git a/src/main/java/io/papermc/paper/event/player/PrePlayerAttackEntityEvent.java b/src/main/java/io/papermc/paper/event/player/PrePlayerAttackEntityEvent.java
index d49999ea597ef247bbf2298e6bf5600bfe003c3e..4ed53b14f94154043e3043f4fef540b45b2b272d 100644
--- a/src/main/java/io/papermc/paper/event/player/PrePlayerAttackEntityEvent.java
+++ b/src/main/java/io/papermc/paper/event/player/PrePlayerAttackEntityEvent.java
@@ -26,6 +26,7 @@ public class PrePlayerAttackEntityEvent extends PlayerEvent implements Cancellab
     private final Entity attacked;
     private boolean cancelled;
     private final boolean willAttack;
+    private boolean forceCrit; // Slice
 
     public PrePlayerAttackEntityEvent(@NotNull Player who, @NotNull Entity attacked, boolean willAttack) {
         super(who);
@@ -55,6 +56,16 @@ public class PrePlayerAttackEntityEvent extends PlayerEvent implements Cancellab
         return this.willAttack;
     }
 
+    // Slice start
+    public boolean isForceCrit() {
+        return forceCrit;
+    }
+
+    public void setForceCrit(boolean forceCrit) {
+        this.forceCrit = forceCrit;
+    }
+    // Slice end
+
     @NotNull
     @Override
     public HandlerList getHandlers() {
