From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Wed, 10 May 2023 08:58:30 -0500
Subject: [PATCH] Add PlayerPreSendChunkPacketEvent


diff --git a/src/main/java/io/papermc/paper/event/player/PlayerPreSendChunkPacketEvent.java b/src/main/java/io/papermc/paper/event/player/PlayerPreSendChunkPacketEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..b91799676d0e178bf974aa52217e551270ba500c
--- /dev/null
+++ b/src/main/java/io/papermc/paper/event/player/PlayerPreSendChunkPacketEvent.java
@@ -0,0 +1,43 @@
+package io.papermc.paper.event.player;
+
+import org.bukkit.Chunk;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class PlayerPreSendChunkPacketEvent extends PlayerEvent {
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+    private final Chunk chunk;
+    private boolean useMode2;
+
+    public PlayerPreSendChunkPacketEvent(@NotNull final Player player, final Chunk chunk) {
+        super(player);
+        this.chunk = chunk;
+    }
+
+    public Chunk getChunk() {
+        return chunk;
+    }
+
+    public boolean isUseMode2() {
+        return useMode2;
+    }
+
+    public void setUseMode2(boolean useMode2) {
+        this.useMode2 = useMode2;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+}
