From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Sun, 10 Sep 2023 07:41:26 -0500
Subject: [PATCH] Add timeDamaged to EntityDamageEvent


diff --git a/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java b/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
index 446b3ffd5caca5344be1c250475679834cd0d4a2..f2beba90acfd251817eb218c35dcf58309d808b1 100644
--- a/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
@@ -30,6 +30,7 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
     private boolean cancelled;
     private final DamageCause cause;
     private final DamageSource damageSource;
+    private long timeDamaged = System.currentTimeMillis(); // Slice
 
     @Deprecated(forRemoval = true)
     public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, final double damage) {
@@ -221,6 +222,22 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
         return damageSource;
     }
 
+    // Slice start
+    /**
+     * @return Get the time the damage event happened.
+     */
+    public long getTimeDamaged() {
+        return timeDamaged;
+    }
+
+    /**
+     * @param timeDamaged Set the time the damage event happened, if you so choose.
+     */
+    public void setTimeDamaged(long timeDamaged) {
+        this.timeDamaged = timeDamaged;
+    }
+    // Slice end
+
     @NotNull
     @Override
     public HandlerList getHandlers() {
