From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Thu, 28 Dec 2023 09:11:49 -0600
Subject: [PATCH] Non-saveable-entities


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 23def071492ccd715693d534cc506936e18f0f46..a1484708c2af2be5c94a0dff1394e8bdf1f90e2e 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1144,4 +1144,19 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     @NotNull String getScoreboardEntryName();
     // Paper end - entity scoreboard name
+
+    // Slice start
+    /**
+     * Returns true if the entity can be saved. If false, the entity will never be serialized or saved.
+     */
+    boolean isSaveable();
+
+    /**
+     * Sets whether the entity can be serialized and saved to disk.
+     *
+     * @param saveable the saveable status
+     * @see #isSaveable()
+     */
+    void setSaveable(boolean saveable);
+    // Slice end
 }
