From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Sun, 7 Jan 2024 08:27:00 -0600
Subject: [PATCH] Vanish


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 8cf34ab37824646cf674237e5dde55459cfbc6f5..0df06f41ca8aef165c5e82c4792f406444c5b218 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1119,4 +1119,19 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     void setSaveable(boolean saveable);
     // Slice end
+
+    // Slice start
+    /**
+     * Returns true if the entity is vanished and cannot emit sounds, effects, etc.
+     */
+    boolean isVanished();
+
+    /**
+     * Sets whether the entity is vanished and cannot emit sounds, effects, etc.
+     *
+     * @param vanished the saveable status
+     * @see #isVanished()
+     */
+    void setVanished(boolean vanished);
+    // Slice end
 }
