From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Wed, 24 Jan 2024 11:39:14 -0600
Subject: [PATCH] PlayerResolveObjectIdEvent


diff --git a/build.gradle.kts b/build.gradle.kts
index 04853c43b99951bf0d4c96ef73724625bdaf018f..504b1b57ba604d261f5f59dc204350f75d1777f8 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -56,6 +56,8 @@ dependencies {
     implementation("org.ow2.asm:asm-commons:9.7")
     // Paper end
 
+    implementation("org.mongodb:bson:4.11.1") // Slice - MongoDB Bson Library for ObjectIds
+
     api("org.apache.maven:maven-resolver-provider:3.9.6") // Paper - make API dependency for Paper Plugins
     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.18")
     compileOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.9.18")
diff --git a/src/main/java/io/papermc/paper/event/player/PlayerResolveObjectIdEvent.java b/src/main/java/io/papermc/paper/event/player/PlayerResolveObjectIdEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..4c28994757812f9c7f59d9eb827908bb64e63118
--- /dev/null
+++ b/src/main/java/io/papermc/paper/event/player/PlayerResolveObjectIdEvent.java
@@ -0,0 +1,44 @@
+package io.papermc.paper.event.player;
+
+import org.bson.types.ObjectId;
+import org.bukkit.Bukkit;
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Objects;
+
+/**
+ * Called when a player trades with a standalone merchant GUI.
+ */
+public class PlayerResolveObjectIdEvent extends PlayerEvent {
+    private static final HandlerList handlers = new HandlerList();
+    private ObjectId objectId;
+
+    public PlayerResolveObjectIdEvent(@NotNull Player player) {
+        super(player, !Bukkit.isPrimaryThread());
+    }
+
+    @Nullable
+    public ObjectId getObjectId() {
+        return this.objectId;
+    }
+
+    public void setObjectId(@NotNull ObjectId objectId) {
+        this.objectId = Objects.requireNonNull(objectId, "ObjectId cannot be null!");
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
+}
diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index 8b0d04d5b39ee817555a36adddc39b18fc6f0d02..6a1df3ccfadfa70594fd71c6d24d94e43297742b 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -22,6 +22,11 @@ import org.jetbrains.annotations.Nullable;
  */
 public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder {
 
+    // Slice start
+    org.bson.types.@Nullable ObjectId getObjectId();
+    void setObjectId(@Nullable org.bson.types.ObjectId objectId);
+    // Slice end
+
     // Paper start
     @Override
     org.bukkit.inventory.@NotNull EntityEquipment getEquipment();
