From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Sun, 27 Feb 2022 09:47:57 -0600
Subject: [PATCH] Add PlayerShieldDisableEvent


diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 850a4f8a9bd1dc30b13205e158fcbfaa872d6157..aa9d0ec59d0bbba073d416cc40f273e7d7c4cd50 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -1609,6 +1609,7 @@ public abstract class Mob extends LivingEntity {
             float f = 0.25F + (float) EnchantmentHelper.getBlockEfficiency(this) * 0.05F;
 
             if (this.random.nextFloat() < f) {
+                if (!new io.papermc.paper.event.player.PlayerShieldDisableEvent((org.bukkit.entity.Player) player.getBukkitEntity(), getBukkitEntity()).callEvent()) return; // Slice
                 player.getCooldowns().addCooldown(Items.SHIELD, 100);
                 this.level.broadcastEntityEvent(player, (byte) 30);
             }
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index d85da02be7f1d494af8f2cc840c49073774f5f9c..2c0c9019245e94a4c89c948eafa837a7d6279b7e 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -933,7 +933,7 @@ public abstract class Player extends LivingEntity {
     protected void blockUsingShield(LivingEntity attacker) {
         super.blockUsingShield(attacker);
         if (attacker.getMainHandItem().getItem() instanceof AxeItem) {
-            this.disableShield(true);
+            this.disableShield(true, attacker); // Slice
         }
 
     }
@@ -1404,6 +1404,12 @@ public abstract class Player extends LivingEntity {
     }
 
     public void disableShield(boolean sprinting) {
+        // Slice start
+        disableShield(sprinting, null);
+    }
+
+    public void disableShield(boolean sprinting, @Nullable LivingEntity attacker) {
+        // Slice end
         float f = 0.25F + (float) EnchantmentHelper.getBlockEfficiency(this) * 0.05F;
 
         if (sprinting) {
@@ -1411,6 +1417,8 @@ public abstract class Player extends LivingEntity {
         }
 
         if (this.random.nextFloat() < f) {
+            org.bukkit.entity.Entity finalAttacker = attacker != null ? attacker.getBukkitEntity() : null; // Slice
+            if (!new io.papermc.paper.event.player.PlayerShieldDisableEvent((org.bukkit.entity.Player) getBukkitEntity(), finalAttacker).callEvent()) return; // Slice
             this.getCooldowns().addCooldown(Items.SHIELD, 100);
             this.stopUsingItem();
             this.level.broadcastEntityEvent(this, (byte) 30);
