From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Mon, 10 Apr 2023 07:31:54 -0500
Subject: [PATCH] Add provided Material to getDrops


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index 7621a8fea46e16bb1aa031ec426721e7d4a4444f..4455e7d6265ba658c24615ec092cc36643fd4cb2 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -616,7 +616,18 @@ public class CraftBlock implements Block {
 
     @Override
     public Collection<ItemStack> getDrops(ItemStack item, Entity entity) {
-        net.minecraft.world.level.block.state.BlockState iblockdata = this.getNMS();
+        return getDrops(null, item, entity); // Slice start
+    }
+
+    @Override
+    public Collection<ItemStack> getDrops(Material blockType, ItemStack item, Entity entity) {
+        net.minecraft.world.level.block.state.BlockState iblockdata;
+        if (blockType == null) {
+            iblockdata = this.getNMS();
+        } else {
+            iblockdata = ((CraftBlockData) blockType.createBlockData()).getState();
+        }
+
         net.minecraft.world.item.ItemStack nms = CraftItemStack.asNMSCopy(item);
 
         // Modelled off EntityHuman#hasBlock
@@ -627,6 +638,7 @@ public class CraftBlock implements Block {
             return Collections.emptyList();
         }
     }
+    // Slice end
 
     @Override
     public boolean isPreferredTool(ItemStack item) {
