From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Mon, 28 Feb 2022 08:56:16 -0600
Subject: [PATCH] Add provided Material to getDrops


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index 1ed2dcb6f28a391689b5dd10073422bae1dcd8bb..f9dbef5ac6ff4cc3b6991b226253e895690a13f5 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -591,7 +591,18 @@ public class CraftBlock implements Block {
 
     @Override
     public Collection<ItemStack> getDrops(ItemStack item, Entity entity) {
-        net.minecraft.world.level.block.state.BlockState iblockdata = this.getNMS();
+        return getDrops(null, item, entity); // Slice start
+    }
+
+    @Override
+    public Collection<ItemStack> getDrops(Material blockType, ItemStack item, Entity entity) {
+        net.minecraft.world.level.block.state.BlockState iblockdata;
+        if (blockType == null) {
+            iblockdata = this.getNMS();
+        } else {
+            iblockdata = ((CraftBlockData) blockType.createBlockData()).getState();
+        }
+
         net.minecraft.world.item.ItemStack nms = CraftItemStack.asNMSCopy(item);
 
         // Modelled off EntityHuman#hasBlock
@@ -602,6 +613,7 @@ public class CraftBlock implements Block {
             return Collections.emptyList();
         }
     }
+    // Slice end
 
     @Override
     public boolean isPreferredTool(ItemStack item) {
