From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Mon, 10 Apr 2023 08:40:20 -0500
Subject: [PATCH] Ignore durability changes for equipment updates


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index f0d565ac0b3c8d3c3a200348a16e7db21ff3920a..f1459e5929f178cdab864506d1a485df61fde035 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3167,7 +3167,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     public boolean equipmentHasChanged(ItemStack stack, ItemStack stack2) {
-        return !ItemStack.matches(stack2, stack);
+        return !ItemStack.isSame(stack2, stack);
     }
 
     private void handleHandSwap(Map<EquipmentSlot, ItemStack> equipmentChanges) {
