From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Mon, 10 Apr 2023 08:47:43 -0500
Subject: [PATCH] Add PlayerPreChunkLoadEvent


diff --git a/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java b/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
index a495d692841c1a6056ca82519ee9b9c66f1ff5cd..18e0fd0bcf3036f986f2726c89592f875a3c630e 100644
--- a/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
+++ b/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
@@ -411,7 +411,7 @@ public class RegionizedPlayerChunkLoader {
         private boolean canGenerateChunks = true;
 
         private final ArrayDeque<ChunkHolderManager.TicketOperation<?, ?>> delayedTicketOps = new ArrayDeque<>();
-        private final LongOpenHashSet sentChunks = new LongOpenHashSet();
+        public final LongOpenHashSet sentChunks = new LongOpenHashSet(); // Slice -> public
 
         private static final byte CHUNK_TICKET_STAGE_NONE           = 0;
         private static final byte CHUNK_TICKET_STAGE_LOADING        = 1;
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 347becd2658a257a7253edccf62fb5923da28b29..524f56735c3f4781be86d0f31e8fb17c578e5b2f 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -531,6 +531,12 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         return true;
     }
 
+    @Override
+    public it.unimi.dsi.fastutil.longs.LongOpenHashSet getSentChunks(Player p) {
+        ServerPlayer craftPlayer = ((CraftPlayer) p).getHandle();
+        return craftPlayer.chunkLoader.sentChunks;
+    }
+
     @Override
     public boolean isChunkInUse(int x, int z) {
         return this.isChunkLoaded(x, z);
