From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Mon, 10 Apr 2023 08:46:54 -0500
Subject: [PATCH] AntiXray Bypass


diff --git a/src/main/java/com/destroystokyo/paper/antixray/ChunkPacketBlockControllerAntiXray.java b/src/main/java/com/destroystokyo/paper/antixray/ChunkPacketBlockControllerAntiXray.java
index e7fe98ea30ae6d0baea3ec1f9f98a89502a49a12..7e2651aa63cc9ddbe28fb9ecbc5188358d52e607 100644
--- a/src/main/java/com/destroystokyo/paper/antixray/ChunkPacketBlockControllerAntiXray.java
+++ b/src/main/java/com/destroystokyo/paper/antixray/ChunkPacketBlockControllerAntiXray.java
@@ -169,6 +169,7 @@ public final class ChunkPacketBlockControllerAntiXray extends ChunkPacketBlockCo
 
     @Override
     public boolean shouldModify(ServerPlayer player, LevelChunk chunk) {
+        if (player.getBukkitEntity().hasAntiXrayBypass()) return false; // Slice
         return !usePermission || !player.getBukkitEntity().hasPermission("paper.antixray.bypass");
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 16734cf6cb5203a9d40be56672a698b761f8ce8a..7af85d92f87a645b35ffcb1ebf934b7fcfe0ec7b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -193,6 +193,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     public org.bukkit.event.player.PlayerResourcePackStatusEvent.Status resourcePackStatus; // Paper - more resource pack API
     private static final boolean DISABLE_CHANNEL_LIMIT = System.getProperty("paper.disableChannelLimit") != null; // Paper - add a flag to disable the channel limit
     private long lastSaveTime; // Paper - getLastPlayed replacement API
+    private boolean antiXrayBypass; // Slice
 
     public CraftPlayer(CraftServer server, ServerPlayer entity) {
         super(server, entity);
@@ -2924,6 +2925,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
     // Paper end
 
+    // Slice start
+    public boolean hasAntiXrayBypass() {
+        return this.antiXrayBypass;
+    }
+
+    @Override
+    public void toggleAntiXrayBypass(boolean bypass) {
+        this.antiXrayBypass = bypass;
+    }
+    // Slice end
+
     @Override
     public void updateCommands() {
         if (this.getHandle().connection == null) return;
