From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Sat, 13 Aug 2022 08:32:55 -0500
Subject: [PATCH] Add PlayerEntityTrackEvent


diff --git a/src/main/java/net/minecraft/server/level/ServerEntity.java b/src/main/java/net/minecraft/server/level/ServerEntity.java
index 201574cfdd6abbbf346c2feb468409028b4ce7f0..95587048f8aa51cd0247d419d5d51f358db360b1 100644
--- a/src/main/java/net/minecraft/server/level/ServerEntity.java
+++ b/src/main/java/net/minecraft/server/level/ServerEntity.java
@@ -354,6 +354,7 @@ public class ServerEntity {
             }
         }
 
+        this.entity.level.getCraftServer().getPluginManager().callEvent(new org.bukkit.event.player.PlayerTrackEntityEvent(entityplayer.getBukkitEntity().getPlayer(), this.entity.getBukkitEntity())); // Slice
     }
 
     private void sendDirtyEntityData() {
