From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Tue, 18 Apr 2023 07:49:59 -0500
Subject: [PATCH] Alllow opening covered chests


diff --git a/src/main/java/net/minecraft/world/level/block/ChestBlock.java b/src/main/java/net/minecraft/world/level/block/ChestBlock.java
index 5e22d175b1048a58802cdf64ac70a8b56329e915..0f15e574b5c0cc8648bc1c0a12f1c72aa8653724 100644
--- a/src/main/java/net/minecraft/world/level/block/ChestBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/ChestBlock.java
@@ -355,9 +355,10 @@ public class ChestBlock extends AbstractChestBlock<ChestBlockEntity> implements
     }
 
     private static boolean isBlockedChestByBlock(BlockGetter world, BlockPos pos) {
-        BlockPos blockposition1 = pos.above();
-
-        return world.getBlockState(blockposition1).isRedstoneConductor(world, blockposition1);
+        return false;
+//        BlockPos blockposition1 = pos.above();
+//
+//        return world.getBlockState(blockposition1).isRedstoneConductor(world, blockposition1);
     }
 
     private static boolean isCatSittingOnChest(LevelAccessor world, BlockPos pos) {
