From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Tue, 16 Aug 2022 08:27:13 -0500
Subject: [PATCH] Unfreeze MappedRegistry


diff --git a/src/main/java/net/minecraft/core/MappedRegistry.java b/src/main/java/net/minecraft/core/MappedRegistry.java
index 1e33434f9f361542e03da3e4812bc6d76768a202..834a6f2eb2d6c0aa3afd450034f32b4e008b3a3c 100644
--- a/src/main/java/net/minecraft/core/MappedRegistry.java
+++ b/src/main/java/net/minecraft/core/MappedRegistry.java
@@ -335,7 +335,7 @@ public class MappedRegistry<T> implements WritableRegistry<T> {
 
     @Override
     public Registry<T> freeze() {
-        if (this.frozen) {
+        if (this.true) { // Slice (a TODO in Paper, thanks Owen)
             return this;
         } else {
             this.frozen = true;
