From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Tue, 18 Apr 2023 07:56:32 -0500
Subject: [PATCH] Disable geode and monsterspawners generation


diff --git a/src/main/java/net/minecraft/data/worldgen/biome/OverworldBiomes.java b/src/main/java/net/minecraft/data/worldgen/biome/OverworldBiomes.java
index 1b6195cd9878e66eb20c45247d7005f9822e79c6..1808399adbbe239c158fc1f2849f5ade8ba1f654 100644
--- a/src/main/java/net/minecraft/data/worldgen/biome/OverworldBiomes.java
+++ b/src/main/java/net/minecraft/data/worldgen/biome/OverworldBiomes.java
@@ -54,8 +54,8 @@ public class OverworldBiomes {
 
     private static void globalOverworldGeneration(BiomeGenerationSettings.Builder generationSettings) {
         BiomeDefaultFeatures.addDefaultCarversAndLakes(generationSettings);
-        BiomeDefaultFeatures.addDefaultCrystalFormations(generationSettings);
-        BiomeDefaultFeatures.addDefaultMonsterRoom(generationSettings);
+//        BiomeDefaultFeatures.addDefaultCrystalFormations(generationSettings); // Slice - Don't generate Geodes in our chunk generation
+//        BiomeDefaultFeatures.addDefaultMonsterRoom(generationSettings);
         BiomeDefaultFeatures.addDefaultUndergroundVariety(generationSettings);
         BiomeDefaultFeatures.addDefaultSprings(generationSettings);
         BiomeDefaultFeatures.addSurfaceFreezing(generationSettings);
