From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Sun, 7 Jan 2024 08:26:59 -0600
Subject: [PATCH] Vanish


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 297e2b8438dde1c1903505e8e7bbc94c3340d2d1..c07e6ff45c2dd51fcd2031bae3d7b32cb614b7ac 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -426,6 +426,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
     public boolean fixedPose = false; // Paper - Expand Pose API
     public boolean saveable = true; // Slice
     public boolean sendMovementPackets = true; // Slice
+    public boolean vanished; // Slice
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
@@ -1325,7 +1326,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
     }
 
     private boolean vibrationAndSoundEffectsFromBlock(BlockPos pos, BlockState state, boolean playSound, boolean emitEvent, Vec3 movement) {
-        if (state.isAir()) {
+        if (vanished || state.isAir()) { // Slice
             return false;
         } else {
             boolean flag2 = this.isStateClimbable(state);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 5db4cf24de9a1c1980f22e72b6f09da146017189..a3ab470b534441c0985e137ea299c8fc46bd982b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1240,4 +1240,16 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         this.entity.saveable = saveable;
     }
     // Slice end
+
+    // Slice start
+    @Override
+    public boolean isVanished() {
+        return this.entity.vanished;
+    }
+
+    @Override
+    public void setVanished(boolean vanished) {
+        this.entity.vanished = vanished;
+    }
+    // Slice end
 }
