From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Sun, 7 Jan 2024 08:26:59 -0600
Subject: [PATCH] Vanish


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index b405ef4ba9343e9dcd7d3b136899bff879e82362..52466d82e07e8c49e462f7b908d6b87cc1ed1dbf 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -426,6 +426,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
     public boolean fixedPose = false; // Paper - Expand Pose API
     public boolean saveable = true; // Slice
     public boolean sendMovementPackets = true; // Slice
+    public boolean vanished; // Slice
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
@@ -1313,7 +1314,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
     }
 
     private boolean vibrationAndSoundEffectsFromBlock(BlockPos pos, BlockState state, boolean playSound, boolean emitEvent, Vec3 movement) {
-        if (state.isAir()) {
+        if (vanished || state.isAir()) { // Slice
             return false;
         } else {
             boolean flag2 = this.isStateClimbable(state);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 3e1c22cd06ecaa26b0c36563a4d0c44ee7e91b1d..df8a79df81afa447121931cd817cff2723e98dcf 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1239,4 +1239,16 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         this.entity.saveable = saveable;
     }
     // Slice end
+
+    // Slice start
+    @Override
+    public boolean isVanished() {
+        return this.entity.vanished;
+    }
+
+    @Override
+    public void setVanished(boolean vanished) {
+        this.entity.vanished = vanished;
+    }
+    // Slice end
 }
