From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: cryptite <cryptite@gmail.com>
Date: Fri, 15 Mar 2024 08:51:06 -0500
Subject: [PATCH] Improve map-saving performance


diff --git a/src/main/java/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java b/src/main/java/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
index 45269115e63cfc3bd7dc740a5694e2cc7c35bcb1..c09e6d89b1595ce26c3489ca9e84a7c2bdf80286 100644
--- a/src/main/java/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
+++ b/src/main/java/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
@@ -331,7 +331,7 @@ public class MapItemSavedData extends SavedData {
             --this.trackedDecorationCount;
         }
 
-        this.setDecorationsDirty();
+        if (mapicon != null) this.setDecorationsDirty(); // Paper - We should not be dirtying this over unless there was mutation.
     }
 
     public static void addTargetDecoration(ItemStack stack, BlockPos pos, String id, MapDecoration.Type type) {
