From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Sun, 7 Jan 2024 08:26:59 -0600
Subject: [PATCH] Vanish


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index cb10e801216d89041c38798be0a2b7f59b73d2fb..91285d21c0f3398a3841d1e17b377845feebd78a 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -424,6 +424,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
     public boolean fixedPose = false; // Paper
     public boolean saveable = true; // Slice
     public boolean sendMovementPackets = true; // Slice
+    public boolean vanished; // Slice
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
@@ -1311,7 +1312,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
     }
 
     private boolean vibrationAndSoundEffectsFromBlock(BlockPos pos, BlockState state, boolean playSound, boolean emitEvent, Vec3 movement) {
-        if (state.isAir()) {
+        if (vanished || state.isAir()) { // Slice
             return false;
         } else {
             boolean flag2 = this.isStateClimbable(state);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 19c56ef79cacfb6d65dea5741d9ce71998c1eacc..c8939926605d6500076c9eae0f0cc008a095ab68 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1239,4 +1239,16 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         this.entity.saveable = saveable;
     }
     // Slice end
+
+    // Slice start
+    @Override
+    public boolean isVanished() {
+        return this.entity.vanished;
+    }
+
+    @Override
+    public void setVanished(boolean vanished) {
+        this.entity.vanished = vanished;
+    }
+    // Slice end
 }
